Fix #7147: Error Flag (ErrorsFound) passed by reference is overriden to false in some functions

jmarrec opened the pull request:
Pull request overview
---------------------
Fix #7147: Error Flag (ErrorsFound) passed by reference is overriden to false in some functions.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

jmarrec committed:
Fix failing tests: now that GetOnlySingleNode wasn't forcing ErrorFound to false, missing avail schedules would make tests throw

jmarrec committed:
Add TODO for another hardsetting the ErrorsFound to false

found while having a watchpoint set on ErrorsFound in DXCoils::GetDXCoils

jmarrec committed:
Add python script to find bool override (and other problems)

jmarrec committed:
Update script to do more, and include a CHECKED_AND_OKED dictionary for false positive

I already started to go through the list of current failures, updating CHECKED_AND_OKED when I found it to be harmless

jmarrec committed:
Found a typo that has been there for at least 19years

https://github.com/NREL/EnergyPlusArchive/blame/95209e0557dc81004125fddafe6a1fc2e2705b7c/SourceCode/SurfaceGeometry.f90#L7747

jmarrec committed:
Keep processing the list, doing some touch-ups where appropriate. From 104 matches down to 35

jmarrec committed:
REVIEW: Interpret section of code in ProcessEPWHeader before refactor so reviewer can see

jmarrec committed:
Cherry pick commit on Ground HX PR

jmarrec committed:
Update GlobalNames::VerifyUniqueADUName to only set ErrorsFound to true

jmarrec committed:
Update GlobalNames::VerifyUniqueBoilerName to only set ErrorsFound to true

jmarrec committed:
Update VerifyUniqueBaseboardName to only set to true

ElectricBaseboardRadiator and SteamBaseboardRadiator.cc were already passing ErrorsFound inside a loop, so it was reset to false everytime, so could lead to testing problems...

jmarrec committed:
VerifyUniqueChillerName

ChillerExhaustAbsorption, ChillerGasAbsorption.cc, could be problematic cause passing a non-dedicated bool in the past

jmarrec committed:
DXCoils.cc, IntegratedHeatPump.cc, VariableSpeedCoils.cc, WaterToAirHeatPump.cc, WaterToAirHeatPumpSimple.cc, could (/will) be problematic cause were passing ErrorsFound in loop...

jmarrec committed:
Misc updates

jmarrec committed:
Remove unused functions and functions without implementation

jmarrec committed:
Change ProcessEPWHeader: see db6ea54b5f8e01f273e7dbef3db367e225455204

jmarrec committed:
Update python script (list of okay'ed methods, and grep in C++ including potential namespaces)

jmarrec committed:
Update python

jmarrec committed:
Set fileencoding to UTF-8 for a few files that weren't (degree sign mostly)

jmarrec committed:
Lint python script

jmarrec committed:
Merge branch 'develop' into PR_opened/Fix_7147_ErrFlagOverride

Resolved 3 conflcts manually

jmarrec committed:
Update python script to find correct source directory path no matter where it's run from

decent_ci runs from root of repo for eg

jmarrec committed:
Fix failing test: these two curves should be defaulted to always 1 per IDD


minor

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_7147_ErrFlagOverride

jmarrec commented:
Note: In 86d0a83 I fixed a failing test thinking it was linked to this PR, but I see that @mjwitte posted a dedicated hotfix in https://github.com/NREL/EnergyPlus/pull/7204 that happens to be better (if something is entered for these curves, and it can't be found, then it'll crash, whereas mine would just let it pass and change it to always 1).

I don't see why his hotfix wouldn't go in develop, so while I'm on it, I've merged it and resolved conflicts

jmarrec committed:
Merge remote-tracking branch 'upstream/HotFixRestoreITEOptionalCurves' into PR_opened/Fix_7147_ErrFlagOverride

Fix merge conflicts

jmarrec commented:
So I fixed the unit tests that failed as a result of this change but I hadn't run the regression/integration ones.... 85 failing :/

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_7147_ErrFlagOverride

jmarrec committed:
Add script to decent_ci

jmarrec commented:
I can't reproduce the failure on Ubuntu locally, and it seems the linux CI doesn't have as many failures in integration tests either...

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into PR_opened/Fix_7147_ErrFlagOverride

jmarrec committed:
Wrap warning in something CI can digest.

Myoldmopar commented on src/EnergyPlus/BranchInputManager.hh:
```diff
@@ -346,7 +349,7 @@ namespace BranchInputManager {
     );
 
-    void TestBranchIntegrity(bool &ErrFound);
+    void TestBranchIntegrity(bool &ErrFound);              // ErrFound is a return value, true or false
```
It seems like there could be a pattern chosen here where if the value of errors_found is used as a pure return value, and intentionally sets it to either true or false, that could just become the actual return value of the function, instead of a reference argument.  Then in cases where it's just set to true, never set to false, then it could be a reference argument.  From the look of the CHECKED_AND_OKED variable, it looks like there may be some 50 functions that are OK'd, and a big chunk of those could be candidates for this conversion.  It would take a little effort and result in more code changes in this PR to do that now, but it would also make the false-positive frequency reduce.

Let me know what you think about it.  I'd be happy to push on that a little bit right now while this PR is open, otherwise it'll just linger and never get done...

jmarrec replied:
I agree that in general if it's a pure return value then the signature should be `bool NameOfFunction()` instead of `void NameOfFunction(bool&)`. It would create somewhat of a shift in typical code patterns in the code base, I don't know if that's acceptable or not, I'll let you be the judge of it.

nrel-bot-2b commented:
@jmarrec it has been 23 days since this pull request was last updated.

Myoldmopar committed:
Merge upstream develop; resolve conflicts

Myoldmopar commented:
@jmarrec I want to get one full set of clean CI results, so I pulled in upstream develop here.  We'll see how it turns out.

Myoldmopar commented:
@jmarrec I'm concerned that the custom_check run hasn't shown results here yet.  I decided to run your script locally.  I get a ton of errors that I am not sure are valid:

```
$ ./scripts/dev/find_byref_bool_override.py 
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 0, "message": "In file DXCoils.cc, line num = 0: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "DXCoils.cc"}
{"tool": "find_byref_bool_overide", "line_num": 0, "message": "In file DXCoils.cc, line num = 0: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "DXCoils.cc"}
...and lots more
```

When I look at it in the debugger, I see that `m.end()` on line 621 is coming up at like character 200, so it's just getting the first couple lines, which puts it inside the license content at the top of the file, and it is causing line_num to be invalid.  I'm not sure all the intent here, so I'm not going to make any changes.  Please advise.

Myoldmopar commented:
OK, so I added some protections to the custom_check processing because it seemed to be failing ungracefully.  I think there's something going on with all the duplicate build results.  The list is supposed to be a unique set, so I'm not sure why it would be an issue, but it does appear to be something related to that.

Anyway, it will likely be cleaned up once my [above comment](https://github.com/NREL/EnergyPlus/pull/7207#issuecomment-495272274) is cleaned up, as the duplicate build warnings won't be present.  But more importantly, it looks like there are failures in the integration tests that need to be investigated.

jmarrec commented:
@Myoldmopar you use python2 I guess. I realize I get different more warnings with python2 **and** a crash, whereas it has one warning in python3 (probably a file that was modified/changed long after I created the PR). I will make it work with python2

Python3:

```
(py37)julien@EnergyPlus (PR_opened/Fix_7147_ErrFlagOverride=)$ python3 scripts/dev/find_byref_bool_override.py 
{"tool": "find_byref_bool_overide", "filename": "HVACControllers.cc", "line": 2254, "messagetype": "error", "message": "Boolean flag `bool &IsConvergedFlag` reset to false in HVACControllers.cc:CheckTempAndHumRatCtrl(), on line 2254"}
```

Python2
```
(py27)julien@EnergyPlus (PR_opened/Fix_7147_ErrFlagOverride=)$ python2 scripts/dev/find_byref_bool_override.py 
{"tool": "find_byref_bool_overide", "line_num": 3, "message": "In file PlantLoopEquip.cc, line num = 3: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantLoopEquip.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file WaterManager.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "WaterManager.cc"}
{"tool": "find_byref_bool_overide", "line_num": 9, "message": "In file PlantCondLoopOperation.cc, line num = 9: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 0, "message": "In file PlantCondLoopOperation.cc, line num = 0: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 6, "message": "In file PlantCondLoopOperation.cc, line num = 6: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 4, "message": "In file PlantCondLoopOperation.cc, line num = 4: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 2, "message": "In file PlantCondLoopOperation.cc, line num = 2: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 4, "message": "In file PlantCondLoopOperation.cc, line num = 4: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 6, "message": "In file PlantCondLoopOperation.cc, line num = 6: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 1, "message": "In file PlantCondLoopOperation.cc, line num = 1: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "PlantCondLoopOperation.cc"}
{"tool": "find_byref_bool_overide", "line_num": 1, "message": "In file OutputReportTabular.cc, line num = 1: \nn_braces is zero which is unexpected.", "messagetype": "warning", "filename": "OutputReportTabular.cc"}
Traceback (most recent call last):
  File "scripts/dev/find_byref_bool_override.py", line 771, in <module>
    all_errors = get_all_errors(source_files)
  File "scripts/dev/find_byref_bool_override.py", line 744, in get_all_errors
    errors = lookup_errors_in_source_file(source_file, found_functions)
  File "scripts/dev/find_byref_bool_override.py", line 647, in lookup_errors_in_source_file
    line = lines[line_num].split('//')[0].strip()
IndexError: list index out of range

```

jmarrec committed:
Fix python2 error: the use of `m` in list comprehension was overriding the other match group m so m.end() was returning the bool one

Python3 doesn't have that problem; scope of m is limited to the list comprehension block

jmarrec committed:
Add a new override for a function

jmarrec commented:
@Myoldmopar interesting problem that was tricky to see initially. If I boil it down to the bare minimum complete and verifiable example (MCVE):

```
In [1]: m = 10                                                                  

In [2]: [m for m in [1, 2, 3]]                                                  
Out[2]: [1, 2, 3]

In [3]: m                                                                       
Out[3]: ???
```

On python2, `Out[3]: m = 3` while on Python3, `Out[3]: m = 10` meaning that in python3 the scope of the `m` variable is limited only to the list comprehension block, while on python2 it overrides the existing `m`.

Python2 is very near its end of life as it won't be maintained past 2020, so I suggest dropping its usage. I would suggest making CMake required at least 3.5 here: https://github.com/NREL/EnergyPlus/blob/develop/CMakeLists.txt#L78

jmarrec commented:
I can't seem to click on the Integration test badges above, I get a 404 error?

jmarrec commented:
I ran the ctest locally on Ubuntu via `ctest -j16 -R "integration.*"`

```
99% tests passed, 2 tests failed out of 660

Total Test time (real) = 2189.10 sec

The following tests FAILED:
	 25 - integration.HospitalLowEnergy (Timeout)
	 27 - integration.VSHeatPumpWaterToAirWithRHControl (Timeout)
Errors while running CTest
```

Running both individually worked, but it took a long time

```
(py37)julien@EnergyPlus-build$ ctest -R "integration.HospitalLowEnergy"
Test project /home/julien/Software/Others/EnergyPlus-build
    Start 25: integration.HospitalLowEnergy
1/1 Test #25: integration.HospitalLowEnergy ....   Passed  944.87 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 945.17 sec
```

```
(py37)julien@EnergyPlus-build$ ctest -R "integration.VSHeatPumpWaterToAirWithRHControl"
Test project /home/julien/Software/Others/EnergyPlus-build
    Start 27: integration.VSHeatPumpWaterToAirWithRHControl
1/1 Test #27: integration.VSHeatPumpWaterToAirWithRHControl ...   Passed  926.06 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 926.36 sec

```

Myoldmopar commented:
The integration test failures I was talking about were failing on [Windows](https://github.com/NREL/EnergyPlus/pull/7207#commitcomment-33654765) not Linux.  So I'm not surprised that they worked fine for you.

If the CI results got cleaned up for some reason and you are getting a 404, just push a commit that merges develop in again and CI will re-run.  We've got enough CI horsepower these days to handle extraneous commits.

Yes, we'll go to Python3 soon, but I'm not going to kill any of our developers builds instantly by changing the Python requirement.  That type of change requires intention and notice, and I'm not willing to make that change as an amendment to a totally unrelated PR.  I suspect we'll make the change after the release of 9.2.  If it's too much trouble to make your script 2/3 compatible, I suggest commenting out the line in the decent yaml file so the script doesn't run at the moment.  But then once we move to Python3, we'll be able to just uncomment the line and use the script then.

jmarrec committed:
Forgot a continue statement

jmarrec commented:
I wasn't suggesting switching to python3 right now, it was just a general comment for the future. I think it makes sense to plan for it and after 9.2 seems like a good idea indeed.
I already made the script python2 compatible (cf https://github.com/NREL/EnergyPlus/pull/7207#issuecomment-496209940) already. 

I cannot merge develop in since I am already caught up with it. I'm building the branch on my other windows machine, so I can look at the Integration test problems locally, and will report back.

jmarrec committed:
Merge branch 'develop' into PR_opened/Fix_7147_ErrFlagOverride

jmarrec committed:
errFlag was initialized to True on the first pass there on Windows

jmarrec committed:
Clean up calls to VerifyUniqueChillerName and remove unecessary flag

jmarrec commented:
https://github.com/NREL/EnergyPlus/pull/7207/commits/96265c932bcbd858dd5bf5cc062bec71ca4ec237 should clean up the windows failures, on my Win10 I get no failures anymore.

I hadn't cleaned up some calls to `VerifyChillerUniqueName`, and the `bool errFlag; declaration;` was making it initialized to `true` on windows (and only windows... hence why you should try to avoid declaring bools without initializing them)

jmarrec commented:
@Myoldmopar all green

Myoldmopar commented:
Yeah, this looks great now!
